File: codebase/core/main/src/AxleModule/Resources/ResourceTemplateProvider.fs
namespace Forest.Resources

open System
open System.Globalization
open System.Reflection
open Axle
open Axle.References
open Axle.Resources
open Forest.Templates.Raw


type [<Sealed>] ResourceTemplateProvider(rm:ResourceManager) =
    [<DefaultValue>]
    val mutable private _bundles : string list voption
    [<DefaultValue>]
    val mutable private _assemblies : Set<string> Nullsafe

    static member BundleName = "ForestTemplates"

    member internal this.AddBundle(bundle : string) = 
        this._bundles <- 
            match this._bundles with
            | ValueSome b -> bundle::b
            | ValueNone -> [bundle]
            |> ValueSome

    member internal this.RegisterAssemblySource(asm : Assembly) =
        let asmname = asm.GetName().Name
        let set =
            match Option.ns2vopt this._assemblies with
            | ValueSome x -> x
            | ValueNone -> Set.empty
        if set |> Set.contains asmname |> not then
            let parseUri = (Axle.Conversion.Parsing.UriParser()).Parse
            for bundle in this.Bundles do
                rm.Bundles
                    .Configure(bundle)
                    .Register(String.Format("assembly://{0}/{1}", asmname, bundle) |> parseUri)
                    |> ignore
            this._assemblies <- set |> Set.add asmname |> Nullsafe.Some

    member internal this.Bundles 
        with get() = 
            match this._bundles with
            | ValueSome b -> b
            | ValueNone -> []
        
    interface ITemplateProvider with
        member this.Load name =
            let rec load bundles =
                match bundles with
                | [] -> raise <| ResourceNotFoundException(name, ResourceTemplateProvider.BundleName, CultureInfo.InvariantCulture)
                | bundle::rest ->
                    let template = rm.Load(bundle, name, CultureInfo.InvariantCulture)
                    if template.HasValue
                    then template.Value.Resolve<Template>()
                    else load rest
            this.Bundles |> load
            

File: codebase/core/main/src/Templates/ResourceTemplateProvider.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Reflection;
using Axle.Resources;

namespace Forest.Templates
{
    [Obsolete("Merge with ForestTemplatesModule class")]
    internal sealed class ResourceTemplateProvider : ITemplateProvider
    {
        internal const string BundleName = "Forest";
        [Obsolete]
        internal const string OldBundleName = "ForestTemplates";

        private readonly ResourceManager _resourceManager;
        private readonly IList<string> _bundles = new List<string>();
        private readonly ISet<string> _assemblies = new HashSet<string>(StringComparer.Ordinal);

        public ResourceTemplateProvider(ResourceManager resourceManager)
        {
            _resourceManager = resourceManager;
        }

        internal void AddBundle(string bundle)
        {
            _bundles.Add(bundle);
        }

        internal void RegisterAssemblySource(Assembly asm)
        {
            var assemblyName = asm.GetName().Name;
            if (_assemblies.Add(assemblyName))
            {
                var uriParser = new Axle.Conversion.Parsing.UriParser();
                foreach (var bundle in _bundles)
                {
                    _resourceManager.Bundles
                        .Configure(bundle)
                        .Register(uriParser.Parse(string.Format("assembly://{0}/{1}", assemblyName, bundle)));
                }
            }
        }

        private Template Load(IList<string> bundles, string name, int index)
        {
            while (true)
            {
                if (index >= bundles.Count)
                {
                    return null;
                }

                var bundle = bundles[index];
                var template = _resourceManager.Load(bundle, name, CultureInfo.InvariantCulture);
                if (template.HasValue)
                {
                    return template.Value.Resolve<Template>();
                }
                index = index + 1;
            }
        }

        public Template Load(string name)
        {
            var result = Load(_bundles, name, 0);
            if (result != null)
            {
                return result;
            }
            throw new ResourceNotFoundException(name, BundleName, CultureInfo.InvariantCulture);
        }
    }
}